proposals/TP2-dynamically-generated-cacheable-xds-resources.md

```diff
@@ -0,0 +1,754 @@
+  by name.
+
+The use-cases that we're aware of for dynamic resource selection have
+an important property that we can take advantage of.  When multiple
+variants of a given resource exist, any given client will only ever use
+one of those variants at a given time.  That means that the parameters
+that affect which variant of the resource is used are required by the
+transport protocol, but they are not required by the client's data model.
+
+It should be noted that caching xDS proxies, unlike "leaf" clients, will
+need to track multiple variants of each resource, since a given caching
+proxy may be serving clients that need different variants of a given
+resource.  However, since caching xDS proxies deal with resources only
+at the transport protocol layer, the resource graph layer is
+essentially irrelevant in that case.
+
+### Dynamic Parameters
+
+With the above property in mind, this document proposes the following
+data structures:
+- **Dynamic parameters**, which are a set of key/value pairs sent by the
+  client when subscribing to a resource.
+- **Dynamic parameter constraints**, which are a set of criteria that
+  can be used to determine whether a set of dynamic parameters matches
+  the constraints.  These constraints are part of the cache key for an
+  xDS resource (in addition to the resource name itself) on xDS servers,
+  xDS clients, and xDS caching proxies.  This provides a mechanism to
+  represent multiple variants of a given resource in a cacheable way.
+
+Both of these data structures are used in the xDS transport protocol,
+but they are not part of the resource name and therefore do not appear as
+part of the resource graph.
+
+When a client subscribes to a resource, it specifies a set of dynamic
+parameters.  In response, the server will send a resource whose dynamic
+parameter constraints match the dynamic parameters in the subscription
+request.  The client will use the dynamic parameter constraints on the
+returned resource to determine which of its subscriptions the resource is
+associated with.
+
+Dynamic parameters, unlike context parameters, will not be
+exact-match-only.  Dynamic parameter constraints will be able to represent
+certain simple types of flexible matching, such as matching an exact
+value or the existance of a key, and simple AND and OR combinations
+of constraints.  This flexible matching semantic means that there may be
+ambiguities when determining which resources match which subscriptions,
+which are discussed below.
+
+#### Constraints Representation
+
+Dynamic parameter constraints will be represented in protobuf form as follows:
+
+```proto
+message DynamicParameterConstraints {
+  // A list of constraints that may be combined with AND or OR semantics.
+  message ConstraintList {
+    // A constraint for a given key.
+    message Constraint {
+      message Exists {}
+      // The key to match against.
+      string key = 1;
+      // How to match.
+      oneof constraint_type {
```

Comment:
If we end up growing this in the future, we will need to teach client, server and intermediate layers about the new proto. Do we need any kind of version negotiation or reporting etc? How would this work in practice?

Reply:
I think it probably makes sense to version the transport protocol itself to support this future proofing. It can be orthogonal to this xRFC but maybe we can add it at the same time as the use of this?

Reply:
@mattklein123 This is the main open issue that I'd like your feedback on.  I think we need to make some trade-offs here between flexibility and complexity, but we may want to err a bit on the side of more flexibility, given that this will be hard to change later.

@htuch I agree that we could address this by versioning the transport protocol.  We've talked before about possibly divorcing the transport protocol versioning from the data model versioning, and I do think that's something we should do.  But we'll need to think carefully about how to do this, and I think it will have to be a separate design.  I don't think we actually need to do it until we actually need to make a change to whatever we put in this proposal, though.